File: Solved/Not_Lisp/fail_permutations.java
import java.util.Stack;
import java.util.ArrayList;
import java.util.Collections;

public class solve {

	static ArrayList<String> permutates = new ArrayList<String>();

	public static void main(String[] args) {
		// https://stackoverflow.com/questions/1235179/simple-way-to-repeat-a-string-in-java
		String brackets = String.join("", 
			Collections.nCopies(10000, "()")); 
				System.out.println("Start permutations" + brackets.length());

		// 10^4 pairs of brackets
		permutation("", brackets);
		System.out.println("Done permutations" + permutates.size());


		String checkBalancedExpr1 = checkBalanced("()()()()()()(");
		System.out.println("a*(b+c)-(d*e) : "+checkBalancedExpr1);
		String checkBalancedExpr2=checkBalanced("(a*(b-c)*{d+e}");
		System.out.println("(a*(b-c)*{d+e} : "+checkBalancedExpr2);
	}

	// https://stackoverflow.com/questions/4240080/generating-all-permutations-of-a-given-string
	private static void permutation(String prefix, String str) {
		int n = str.length();
		if (n == 0)
			permutates.add(prefix);
		else {
			for (int i = 0; i < n; i++)
				permutation(prefix + str.charAt(i), str.substring(0, i) + str.substring(i+1, n));
		}
	}

	public static String checkBalanced(String expr) {
		if (expr.isEmpty())
			return "Balanced";
	 
		Stack<Character> stack = new Stack<Character>();
		for (int i = 0; i < expr.length(); i++) {
			char current = expr.charAt(i);
			if (current == '{' || current == '(' || current == '[' || current == '1') {
				stack.push(current);
			}

			if (current == '}' || current == ')' || current == ']' || current == '0') {
				if (stack.isEmpty())
					return "Not Balanced";
				char last = stack.peek();
				if (current == '}' && last == '{' 
					|| current == ')' && last == '(' 
					|| current == ']' && last == '['
					|| current == '0' && last == '1')
					stack.pop();
				else 
					return "Not Balanced";
			}
		}
	return stack.isEmpty() ? "Balanced" : "Not Balanced";
	}
}

File: Solved/Not_Lisp/fail_permutations.py
class Stack:
     def __init__(self):
         self.items = []

     def isEmpty(self):
         return self.items == []

     def push(self, item):
         self.items.append(item)

     def pop(self):
         return self.items.pop()

     def peek(self):
         return self.items[len(self.items)-1]

     def size(self):
         return len(self.items)

def is_left_bracket(current):
    return current == '1' or current == '{' or current == '(' or current == '[' 

def is_right_bracket(current):
    return current == '0' or current == '}' or current == ')' or current == ']'

def is_pair_bracket(current, last):
    return (current == '0' and last == '1'
                or current == '}' and last == '{' 
                or current == ')' and last == '(' 
                or current == ']' and last == '[')

def check_balanced(str):
    if not str:
        return True

    if str.count('0') != str.count('1'):
        return False

    stack = Stack()
    for current in str:
        #current = str[i]
        if (is_left_bracket(current)):
            stack.push(current)

        if (is_right_bracket(current)):
            if stack.isEmpty():
                return False

            last = stack.peek();
            if (is_pair_bracket(current, last)):
                stack.pop();
            else:
                return False

    return stack.isEmpty()

def get_possibilities(pairs):
    # expressed in binary where 1=( and 0=)
    balanced_count = 0
    mynumb = 2**(pairs*2)
    while mynumb > 0:
        binary = bin(mynumb)[2:] # remove 0b prefix
        if check_balanced(binary):
            balanced_count += 1
        mynumb -=1
    return balanced_count

for x in range(100):
    print(f"{x} -> {get_possibilities(x)}")


    
'''


()()()()
(()()())
((())())
(()(()))
(())(())
((()()))
(((())))


{}<>[]() / {}<>[()]
{<>}[]() / {<>}[()]
{<>[]}() / {<[]>}()
{<>[]()} / {<[]>()} / {<[]()>}

n = 4
p = 3


3 possible 
'''

'''
public static String checkBalanced(String expr) {
    if (expr.isEmpty())
        return "Balanced";
 
    Stack<Character> stack = new Stack<Character>();
    for (int i = 0; i < expr.length(); i++) {
        char current = expr.charAt(i);
        if (current == '{' || current == '(' || current == '[' || current == '1') {
            stack.push(current);
        }

        
    }
return stack.isEmpty() ? "Balanced" : "Not Balanced";
}
'''